### PR TITLE
[Clang][SVE2p1]Add svboolx2 and svboolx4 types for svcreate, svget, s…

### PR DESCRIPTION
**Pull request by CarolineConcatto**
…vset

According to the PR#257[1]

[1]ARM-software/acle#257

Co-authored by: Matthew Devereau <matthew.devereau@arm.com>

### CODE REVIEW

File: clang/include/clang/Basic/arm_sve.td [hunk 1]
```diff
@@ -1316,6 +1321,13 @@ def SVSET_3_BF16 : SInst<"svset3[_{d}]", "33id", "b", MergeNone, "", [IsTupleSet
 def SVSET_4_BF16 : SInst<"svset4[_{d}]", "44id", "b", MergeNone, "", [IsTupleSet], [ImmCheck<1, ImmCheck0_3>]>;
 }
 
+let TargetGuard = "sve2p1" in {
+  def SVGET_2_B : SInst<"svget2[_{d}]", "d2i", "Pc", MergeNone, "", [IsTupleGet], [ImmCheck<1, ImmCheck0_1>]>;
+  def SVGET_4_B : SInst<"svget4[_{d}]", "d4i", "Pc", MergeNone, "", [IsTupleGet], [ImmCheck<1, ImmCheck0_3>]>;
+
+  def SVSET_2_B : SInst<"svset2[_{d}]", "22id", "Pc", MergeNone, "", [IsTupleSet], [ImmCheck<1, ImmCheck0_1>]>;
+  def SVSET_4_B : SInst<"svset4[_{d}]", "44id", "Pc", MergeNone, "", [IsTupleSet], [ImmCheck<1, ImmCheck0_3>]>;
```
Comment (kmclaughlin-arm) [inline clang/include/clang/Basic/arm_sve.td:1329]: Can you please add some tests for the ImmChecks used by svget/svset to acle_sve2p1_imm.cpp?

**Comment by CarolineConcatto** [comment]
I added the tests for svget and svset. I am not sure about the flag yet.

File: clang/test/Sema/aarch64-sve2p1-intrinsics/acle_sve2p1_imm.cpp [hunk 2]
```diff
@@ -167,3 +167,23 @@ void test_svpmov_lane(){
...
+
+  svset2(tuple2, -1, res); // expected-error {{argument value 18446744073709551615 is outside the valid range [0, 1]}}
+  svset2(tuple2, 2,  res); // expected-error {{argument value 2 is outside the valid range [0, 1]}}
+  svset4(tuple4, -1, res); // expected-error {{argument value 18446744073709551615 is outside the valid range [0, 3]}}
+  svset4(tuple4, 4,  res); // expected-error {{argument value 4 is outside the valid range [0, 3]}}
+
+  res = svget2(tuple2, -1); // expected-error {{argument value 18446744073709551615 is outside the valid range [0, 1]}}
+  res = svget2(tuple2,  2); // expected-error {{argument value 2 is outside the valid range [0, 1]}}
+  res = svget4(tuple4, -1); // expected-error {{argument value 18446744073709551615 is outside the valid range [0, 3]}}
+  res = svget4(tuple4,  4); // expected-error {{argument value 4 is outside the valid range [0, 3]}}
```
Comment (kmclaughlin-arm) [inline clang/test/Sema/aarch64-sve2p1-intrinsics/acle_sve2p1_imm.cpp:183]: I don't think you need to use `res =` here, you can just call svget similar to svset above

File: clang/test/Sema/aarch64-sve2p1-intrinsics/acle_sve2p1_imm.cpp [hunk 3]
```diff
@@ -167,3 +167,23 @@ void test_svpmov_lane(){
   zn_u32 = svpmov_lane_u32_m(zn_u32, pn, 5); // expected-error {{argument value 5 is outside the valid range [1, 3]}}
   zn_u64 = svpmov_lane_u64_m(zn_u64, pn, 8); // expected-error {{argument value 8 is outside the valid range [1, 7]}}
 }
+
+__attribute__((target("+sve2p1")))
+void test_svget_b(uint64_t idx, svboolx2_t tuple2, svboolx4_t tuple4){
+  svbool_t res;
```
Comment (kmclaughlin-arm) [inline clang/test/Sema/aarch64-sve2p1-intrinsics/acle_sve2p1_imm.cpp:173]: nit: can this be added to the list of args for `test_svget_b`, along with idx & tuple2/4?

File: clang/include/clang/Basic/arm_sve.td [hunk 4]
```diff
@@ -1316,6 +1321,13 @@ def SVSET_3_BF16 : SInst<"svset3[_{d}]", "33id", "b", MergeNone, "", [IsTupleSet
...
+}
 ////////////////////////////////////////////////////////////////////////////////
```
[the remark this answers is not in the document]
Comment (kmclaughlin-arm) [reply clang/include/clang/Basic/arm_sve.td:1331]: You're right, we only need to add the flag once it's enabled for SME2

**Review by kmclaughlin-arm — COMMENTED** [review]
Thank you for adding the tests @CarolineConcatto!

**Review by kmclaughlin-arm — APPROVED** [review]

**Pull request merged**